File: risk/engine/reinforcements.py
"""Reinforcement calculation for Risk game."""

from risk.engine.map_graph import MapGraph
from risk.models.game_state import GameState


def calculate_reinforcements(
    state: GameState, map_graph: MapGraph, player_index: int
) -> int:
    """Calculate total reinforcement armies for a player.

    Base: max(territories_owned // 3, 3)
    Plus: bonus for each fully controlled continent.
    """
    player_territories = {
        name
        for name, ts in state.territories.items()
        if ts.owner == player_index
    }

    # Base reinforcements: territory count / 3, minimum 3
    base = max(len(player_territories) // 3, 3)

    # Continent bonuses
    bonus = 0
    for continent in map_graph._continent_bonuses:
        if map_graph.controls_continent(continent, player_territories):
            bonus += map_graph.continent_bonus(continent)

    return base + bonus

File: mobile/lib/engine/reinforcements.dart
/// Reinforcement calculation for Risk game.
/// Pure Dart port of risk/engine/reinforcements.py.
/// Zero Flutter imports — runs in isolates and pure-Dart tests.

import 'dart:math';
import 'map_graph.dart';
import 'models/game_state.dart';

/// Calculate total reinforcement armies for a player.
///
/// Base: max(territories_owned ~/ 3, 3)
/// Plus: bonus for each fully controlled continent.
int calculateReinforcements(
  GameState state,
  MapGraph mapGraph,
  int playerIndex,
) {
  final playerTerritories = state.territories.entries
      .where((e) => e.value.owner == playerIndex)
      .map((e) => e.key)
      .toSet();

  // Base reinforcements: territory count / 3, minimum 3 (integer division)
  final base = max(playerTerritories.length ~/ 3, 3);

  // Continent bonuses
  int bonus = 0;
  for (final continent in mapGraph.continentNames) {
    if (mapGraph.controlsContinent(continent, playerTerritories)) {
      bonus += mapGraph.continentBonus(continent);
    }
  }

  return base + bonus;
}
